2.5-huffman.ipynb

```diff
@@ -1,25 +1,248 @@
     "\n",
-    "Huffman encoding can optimize the TapTree structure, and thus potentially lower the size of revealed branches, by taking into account the estimated frequency by which each TapLeaf will occur. Since some spending scenarios are more likely to occur than others, ideally we would optimize the tree structure such that those spends appear closer to the root of the tree, and therefore require a smaller merkle proof when spending.\n",
+    "We reconstruct the same TapTree from the example above using the `TapTree.huffman_constructor()` method. \n",
+    "\n",
+    "* That function takes a list of `(weight,tapleaf)` tuples, where:\n",
+    "    * `weight` is an `int` \n",
+    "    * `tapleaf` is a `TapLeaf` object\n",
+    "   "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Example 2.5.2: Construct a TapTree with the Huffman constructor method\n",
+    "* We compare the resulting TapTree with the one constructed manually in example 2.5.1"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "taptree2 = TapTree()\n",
+    "taptree2.key = internal_pubkey\n",
+    "taptree2.huffman_constructor([(5, tapleafA), (4, tapleafB), (3, tapleafC), (2, tapleafD), (1, tapleafE)])\n",
+    "print(\"Taptree descriptor: {}\\n\".format(taptree2.desc))\n",
     "\n",
-    "We reconstruct the same taptree from the above examples using the `huffman_constructor()`. That function takes a list of `(weight,tapleaf)` tuples, where `weight` is an `int` and `tapleaf` is a `TapLeaf` object. A higher weight:\n",
-    "   \n",
-    "* Indicates a higher likelihood of execution\n",
-    "* Means the script will be placed closer to the root if possible\n",
-    "* Results in a smaller inclusion proof and lower spending fees"
+    "segwit_v1_script2, tweak2, control_map2 = taptree2.construct()\n",
+    "print(\"TapTrees are identical: {}\".format(tweak == tweak2))"
    ]
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "![test](images/taptree3.jpg)"
+    "#### _Programming Exercise 2.5.3:_ Determine the Probability Weights for this Tree Structure\n",
+    "\n",
+    "Given the TapTree structure shown below, determine a valid set of relative probability weights for the Huffman constructor. In other words, try to follow the Huffman algorithm from the lowest depth of the tree and determine what probability weights would be necessary in order to construct the tree depicted in the image below.\n",
+    "\n",
+    "For TapLeafs A through F:\n",
+    "* Generate `pk` TapScripts for each.\n",
+    "* Assign their probability weights so that the TapTree resulting from the Huffman constructor has the desired tree structure."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "![test](images/huffman_intro1.jpg)"
+   ]
+  },
+  {
+   "cell_type": "code",
```

Comment:
This exercise would be more effective if there was a clear pass/fail test at the bottom. I just entered random weights and there's no way for me to know whether I got the right answer or now.

Reply:
Yes I agree. I have added a pass/fail test at the bottom which prints out heights of constructed tree. Cheers.